modules/finality-verifier/src/lib.rs

```diff
@@ -175,32 +322,207 @@ pub mod pallet {
+		// "travelling back in time" (which would be indicative of something bad, e.g a hard-fork).
+		ensure!(best_finalized.number() < header.number(), <Error<T>>::ConflictingFork);
+
+		// We don't support forced changes - at that point governance intervention is required.
+		ensure!(
+			super::find_forced_change(&header).is_none(),
+			<Error<T>>::UnsupportedScheduledChange
+		);
+
+		if let Some(change) = super::find_scheduled_change(&header) {
+			// GRANDPA only includes a `delay` for forced changes, so this isn't valid.
+			ensure!(change.delay == Zero::zero(), <Error<T>>::UnsupportedScheduledChange);
+
+			let next_authorities = bp_header_chain::AuthoritySet {
+				authorities: change.next_authorities,
+				set_id: <CurrentAuthoritySet<T>>::get().set_id + 1,
+			};
+
+			// Since our header schedules a change and we know the delay is 0, it must also enact
+			// the change.
+			<CurrentAuthoritySet<T>>::put(next_authorities);
+		};
+
+		<BestFinalized<T>>::put(header.hash());
+		<ImportedHeaders<T>>::insert(header.hash(), header);
+
+		Ok(())
+	}
+
+	/// Since this writes to storage with no real checks this should only be used in functions that
+	/// were called by a trusted origin.
+	fn initialize_bridge<T: Config>(init_params: super::InitializationData<BridgedHeader<T>>) {
+		let super::InitializationData {
+			header,
+			authority_list,
+			set_id,
+			is_halted,
+		} = init_params;
+
+		let initial_hash = header.hash();
+		<InitialHash<T>>::put(initial_hash);
+		<BestFinalized<T>>::put(initial_hash);
+		<ImportedHeaders<T>>::insert(initial_hash, header);
+
+		let authority_set = bp_header_chain::AuthoritySet::new(authority_list, set_id);
+		<CurrentAuthoritySet<T>>::put(authority_set);
+
+		<IsHalted<T>>::put(is_halted);
+	}
+
+	/// Ensure that the origin is either root, or `ModuleOwner`.
+	fn ensure_owner_or_root<T: Config>(origin: T::Origin) -> Result<(), BadOrigin> {
+		match origin.into() {
+			Ok(RawOrigin::Root) => Ok(()),
+			Ok(RawOrigin::Signed(ref signer)) if Some(signer) == <ModuleOwner<T>>::get().as_ref() => Ok(()),
+			_ => Err(BadOrigin),
+		}
+	}
+
+	/// Ensure that the pallet is in operational mode (not halted).
+	fn ensure_operational<T: Config>() -> Result<(), Error<T>> {
+		if <IsHalted<T>>::get() {
+			Err(<Error<T>>::Halted)
+		} else {
+			Ok(())
+		}
 	}
 }
 
+impl<T: Config> Pallet<T> {
+	/// Get the best finalized header the pallet knows of.
+	///
+	/// Returns a dummy header if there is no best header. This can only happen
+	/// if the pallet has not been initialized yet.
+	pub fn best_finalized() -> BridgedHeader<T> {
+		let hash = <BestFinalized<T>>::get();
+		<ImportedHeaders<T>>::get(hash).unwrap_or_else(|| {
+			<BridgedHeader<T>>::new(
+				Default::default(),
+				Default::default(),
+				Default::default(),
+				Default::default(),
+				Default::default(),
+			)
+		})
+	}
+
+	/// Check if a particular header is known to the bridge pallet.
+	pub fn is_known_header(hash: BridgedBlockHash<T>) -> bool {
```

Comment:
I don't think this is super useful anymore for syncing since it looks like relayers only care about the best finalized headers. Do you think this is helpful to keep @svyatonik?